File: berty-bridge-expo/ios/BertyBridgeExpoModule.swift
import ExpoModulesCore
import Bertybridge

public class BertyBridgeExpoModule: Module {
    let logger = BertyLogger("tech.berty.react")

    // protocol
    var bridgeMessenger: BertybridgeBridge?
    var remoteBridge: BertybridgeRemoteBridge?
    var serviceClient: BertybridgeServiceClientProtocol?
    var connectivityDriver = ConnectivityDriver()
    var appRootDir: String?
    var sharedRootDir: String?
    var debug: Bool?
    
    enum logLevel: String, Enumerable {
        case debug
        case info
        case warn
        case error
    }
    
    struct logOpts: Record {
        @Field
        var level: logLevel?
        
        @Field
        var message: String?
    }
    
  public func definition() -> ModuleDefinition {
    Name("BertyBridgeExpo")

    OnCreate {
      // set berty dir for persistence
      self.sharedRootDir = try! RootDirGet()

      let docDir = try! FileManager.default.url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
      self.appRootDir = docDir.appendingPathComponent("berty", isDirectory: true).path
        
#if DEBUG_LOGS
      debug = true;
#else
      debug = false;
#endif
    }
      
    OnDestroy {
      do {
          if self.bridgeMessenger != nil {
              try self.bridgeMessenger?.close()
              self.bridgeMessenger = nil
          }
          if self.remoteBridge != nil {
              try self.remoteBridge?.close()
              self.remoteBridge = nil
          }
          self.serviceClient = nil
      } catch let error as NSError {
          self.logger.error("\(String(describing: error.code))")
      }
    }
    
    Constants([
        "debug": self.debug
    ])
      
    AsyncFunction("initBridge") { (promise: Promise) in
        do {
            if self.bridgeMessenger != nil || self.remoteBridge != nil || self.serviceClient != nil {
                throw NSError(domain: "tech.berty.gobridge", code: 1, userInfo: [NSLocalizedDescriptionKey : "already started"])
            }

            var err: NSError?
            guard let config = BertybridgeNewBridgeConfig() else {
                throw NSError(domain: "tech.berty.gobridge", code: 2, userInfo: [NSLocalizedDescriptionKey : "unable to create config"])
            }

            // get user preferred languages
            let preferredLanguages: String = Locale.preferredLanguages.joined(separator: ",")

            config.setLifeCycleDriver(LifeCycleDriver.shared)
            config.setNotificationDriver(NotificationDriver.shared)
            config.setKeystoreDriver(KeystoreDriver.shared)
            config.setPreferredLanguages(preferredLanguages)

            // @TODO(gfanton): make this dir in golang
            var isDirectory: ObjCBool = true
            var exist = FileManager.default.fileExists(atPath: self.sharedRootDir!, isDirectory: &isDirectory)
            if !exist {
                try FileManager.default.createDirectory(atPath: self.sharedRootDir!, withIntermediateDirectories: true, attributes: nil)
            }
          
            exist = FileManager.default.fileExists(atPath: self.appRootDir!, isDirectory: &isDirectory)
            if !exist {
                try FileManager.default.createDirectory(atPath: self.appRootDir!, withIntermediateDirectories: true, attributes: nil)
            }

            // Disable iOS backup
            var values = URLResourceValues()
            values.isExcludedFromBackup = true
          
            var appRootDirURL = URL(fileURLWithPath: self.appRootDir!)
            try appRootDirURL.setResourceValues(values)

            var sharedRootDirURL = URL(fileURLWithPath: self.sharedRootDir!)
            try sharedRootDirURL.setResourceValues(values)

            // Set root directories
            config.setAppRootDir(self.appRootDir)
            config.setSharedRootDir(self.sharedRootDir)

            config.setConnectivityDriver(self.connectivityDriver);

            let bridgeMessenger = BertybridgeNewBridge(config, &err)
            if err != nil {
                throw err!
            }

            self.bridgeMessenger = bridgeMessenger
            self.serviceClient = bridgeMessenger // bridgeMessenger implements ServiceClient interface

            BertyLogger.useBridge(self.bridgeMessenger)

            promise.resolve(true)
        } catch let error as NSError {
            promise.reject(error)
        }
    }

      AsyncFunction("initBridgeRemote") { (address: String, promise: Promise) in
          do {
              if self.remoteBridge != nil || self.bridgeMessenger != nil || self.serviceClient != nil {
                  throw NSError(domain: "tech.berty.gobridge", code: 1, userInfo: [NSLocalizedDescriptionKey : "already started"])
              }

              var err: NSError?
              guard let config = BertybridgeNewRemoteBridgeConfig() else {
                  throw NSError(domain: "tech.berty.gobridge", code: 2, userInfo: [NSLocalizedDescriptionKey : "unable to create config"])
              }

              // Disable iOS backup
              var values = URLResourceValues()
              values.isExcludedFromBackup = true

              let remoteBridge = BertybridgeNewRemoteBridge(address, config, &err)
              if err != nil {
                  throw err!
              }

              self.remoteBridge = remoteBridge
              self.serviceClient = remoteBridge // remoteBridge implements ServiceClient interface

              promise.resolve(true)
          } catch let error as NSError {
              promise.reject(error)
          }
      }

      AsyncFunction("connectService") { (serviceName: String, address: String, promise: Promise) in
          do {
              guard let remoteBridge = self.remoteBridge else {
                  throw NSError(domain: "tech.berty.gobridge", code: 4, userInfo: [NSLocalizedDescriptionKey : "remoteBridge isn't started"])
              }
              try remoteBridge.connectService(serviceName, address: address)
              promise.resolve(true)
          } catch let error as NSError {
              promise.reject(error)
          }
      }

      AsyncFunction("closeBridge") { (promise: Promise) in
            do {
                if self.bridgeMessenger != nil {
                    try self.bridgeMessenger?.close()
                    self.bridgeMessenger = nil
                }
                if self.remoteBridge != nil {
                    try self.remoteBridge?.close()
                    self.remoteBridge = nil
                }
                self.serviceClient = nil
                promise.resolve(true)
            } catch let error as NSError {
                promise.reject(error)
            }
        }

      AsyncFunction("invokeBridgeMethod") { (method: String, b64message: String, promise: Promise) in
            do {
                guard let serviceClient = self.serviceClient else {
                    throw NSError(domain: "tech.berty.gobridge", code: 3, userInfo: [NSLocalizedDescriptionKey : "serviceClient isn't started"])
                }

                let promise = PromiseBlock(promise)
                serviceClient.invokeBridgeMethod(with: promise, method: method, b64message: b64message)
            } catch let error as NSError {
                promise.reject(error)
            }
        }

      AsyncFunction("getProtocolAddr") { (promise: Promise) in
            do {
                if self.bridgeMessenger == nil {
                    throw NSError(domain: "tech.berty.gobridge", code: 4, userInfo: [NSLocalizedDescriptionKey : "bridgeMessenger isn't started"])
                }

                let addr: [String] = []
                promise.resolve(addr)
            } catch let error as NSError {
                promise.reject(error)
            }
        }
      
      AsyncFunction("clearStorage") { (promise: Promise) in
          do {
              if FileManager.default.fileExists(atPath: self.appRootDir!) {
                  try FileManager.default.removeItem(atPath: self.appRootDir!)
              }
              if FileManager.default.fileExists(atPath: self.sharedRootDir!) {
                  try FileManager.default.removeItem(atPath: self.sharedRootDir!)
              }
              promise.resolve(true)
          }
          catch let error as NSError {
              promise.reject(error)
          }
      }

      Function("log") { (opts: logOpts) in
          #if !CFG_APPSTORE
          if let message = opts.message {
              let level = opts.level ?? logLevel.info

              self.logger.log(BertyLogger.LogLevel(rawValue: level.rawValue.uppercased()) ?? .INFO, message)
          }
          #endif
      }

    // Enables the module to be used as a native view. Definition components that are accepted as part of the
    // view definition: Prop, Events.
    View(BertyBridgeExpoView.self) {
      // Defines a setter for the `name` prop.
      Prop("name") { (view: BertyBridgeExpoView, prop: String) in
        print(prop)
      }
    }
  }
}

File: berty-bridge-expo/android/src/main/java/tech/berty/bertybridgeexpo/BertyBridgeExpoModule.kt
package tech.berty.bertybridgeexpo

import android.bluetooth.BluetoothAdapter
import android.content.Context
import android.content.IntentFilter
import android.content.res.Resources
import android.net.ConnectivityManager
import android.os.Build
import android.provider.CalendarContract.Instances.EVENT_ID
import android.util.Log
import bertybridge.Bertybridge
import bertybridge.Bridge
import bertybridge.PromiseBlock
import bertybridge.RemoteBridge
import bertybridge.ServiceClient
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import expo.modules.kotlin.types.Enumerable
import tech.berty.bertybridgeexpo.addressbook.AddressBook
import tech.berty.bertybridgeexpo.gobridge.ConnectivityDriver
import tech.berty.bertybridgeexpo.gobridge.GoBridgeAlreadyStartedError
import tech.berty.bertybridgeexpo.gobridge.GoBridgeConfigError
import tech.berty.bertybridgeexpo.gobridge.GoBridgeKeystoreNotStartedError
import tech.berty.bertybridgeexpo.gobridge.GoBridgeNotStartedError
import tech.berty.bertybridgeexpo.gobridge.GoBridgeStorageError
import tech.berty.bertybridgeexpo.gobridge.GoBridgeTempDirError
import tech.berty.bertybridgeexpo.gobridge.KeystoreDriver
import tech.berty.bertybridgeexpo.gobridge.KotlinPromiseBlock
import tech.berty.bertybridgeexpo.gobridge.Logger
import tech.berty.bertybridgeexpo.gobridge.MDNSLockerDriver
import tech.berty.bertybridgeexpo.gobridge.NetDriver
import tech.berty.bertybridgeexpo.gobridge.bledriver.BleInterface
import tech.berty.bertybridgeexpo.notification.NotificationService
import tech.berty.bertybridgeexpo.rootdir.RootDirModule
import java.io.File

enum class logLevel(val value: String) : Enumerable {
    debug("debug"),
    info("info"),
    warn("warn"),
    error("error"),
}

class logOpts : Record {
    @Field
    val level: logLevel? = null

    @Field
    val message: String? = null
}

class BertyBridgeExpoModule : Module() {
    private val TAG: String = "GoBridge"

    // protocol
    private var remoteBridge: RemoteBridge? = null
    private var serviceClient: ServiceClient? = null
    private var rootDir: File? = null
    private var tempDir: File? = null
    private var keystoreDriver: KeystoreDriver? = null
    private var addressBook: AddressBook? = null
    private var notificationService: NotificationService? = null

    // Each module class must implement the definition function. The definition consists of components
    // that describes the module's functionality and behavior.
    // See https://docs.expo.dev/modules/module-api for more details about available components.
    override fun definition() = ModuleDefinition {
        // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
        // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
        // The module will be accessible from `requireNativeModule('BertyBridgeExpo')` in JavaScript.
        Name("BertyBridgeExpo")

        // Sets constant properties on the module. Can take a dictionary or a closure that returns a dictionary.
        Constants(
            "PI" to Math.PI
        )

        Events(EVENT_ID)

        OnCreate {
            val reactContext = appContext.reactContext!!

            try {
                keystoreDriver = KeystoreDriver(reactContext)
            } catch (e: Exception) {
                e.printStackTrace()
            }

            rootDir = File(RootDirModule(reactContext).getRootDir())
            tempDir = File(reactContext.cacheDir.absolutePath + "/berty")
            addressBook = AddressBook(reactContext)
            notificationService = NotificationService(this@BertyBridgeExpoModule)
        }

        OnDestroy {
            try {
                if (bridgeMessenger != null) {
                    bridgeMessenger!!.close()
                }
                if (remoteBridge != null) {
                    remoteBridge!!.close()
                }
            } catch (e: CodedException) {
                Logger.e(TAG, "bridge close error", e)
            }
            bridgeMessenger = null
            remoteBridge = null
            serviceClient = null
        }

        AsyncFunction("initBridge") { promise: Promise ->
            Log.i("initBridge", "initBridge called")
            try {
                if (bridgeMessenger != null || remoteBridge != null || serviceClient != null) {
                    throw GoBridgeAlreadyStartedError()
                }

                if (keystoreDriver == null) {
                    throw GoBridgeKeystoreNotStartedError()
                }

                val reactContext = appContext.reactContext!!

                val config = Bertybridge.newBridgeConfig() ?: throw GoBridgeConfigError()

                // set net driver
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                    val inet = NetDriver()
                    config.setNetDriver(inet)
                }

                // set mdns locker driver
                val imdnslocker = MDNSLockerDriver(reactContext)
                config.setMDNSLocker(imdnslocker)

                // load and set user preferred language
                var tags: String? = null
                tags = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    Resources.getSystem().configuration.locales.toLanguageTags()
                } else {
                    Resources.getSystem().configuration.locale.toLanguageTag()
                }
                config.setPreferredLanguages(tags)

                // set root dir
                config.setAppRootDir(rootDir!!.absolutePath)

                // set temp dir
                if (tempDir != null && !tempDir!!.exists()) {
                    if (!tempDir!!.mkdirs()) {
                        throw GoBridgeTempDirError()
                    }
                }
                config.setAndroidCacheDir(tempDir!!.absolutePath)

                // set ble driver
                val bleDriver: BleInterface = BleInterface(reactContext, true)
                config.setBleDriver(bleDriver)

                // set NearBy driver
                // BertyNearbyDriver NBDriver = new BertyNearbyDriver(reactContext);
                // config.setNBDriver(NBDriver);

                // set native keystore driver
                config.setKeystoreDriver(keystoreDriver)

                val connectivityDriver = ConnectivityDriver(reactContext!!)
                reactContext!!.registerReceiver(
                    connectivityDriver,
                    IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION)
                )
                reactContext!!.registerReceiver(
                    connectivityDriver,
                    IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED)
                )
                config.setConnectivityDriver(connectivityDriver)

                bridgeMessenger = Bertybridge.newBridge(config)
                serviceClient =
                    bridgeMessenger // bridgeMessenger implements ServiceClient interface

                Logger.useBridge(bridgeMessenger)

                promise.resolve(true)
            } catch (err: CodedException) {
                promise.reject(err)
            }
        }

        Function("log") { opts: logOpts ->
            if (opts.message != null) {
                val level = opts.level ?: "INFO"

                // log
                Logger.log(
                    Logger.LogLevel.fromString(level.toString()),
                    "react-native",
                    opts.message
                )
            }
        }

        AsyncFunction("initBridgeRemote") { address: String, promise: Promise ->
            try {
                if (remoteBridge != null || bridgeMessenger != null || serviceClient != null) {
                    throw GoBridgeAlreadyStartedError()
                }

                if (keystoreDriver == null) {
                    throw GoBridgeKeystoreNotStartedError()
                }

                val config = Bertybridge.newRemoteBridgeConfig() ?: throw GoBridgeConfigError()

                remoteBridge = Bertybridge.newRemoteBridge(address, config)
                serviceClient = remoteBridge // remoteBridge implements ServiceClient interface
                promise.resolve(true)
            } catch (err: CodedException) {
                promise.reject(err)
            }
        }

        AsyncFunction("connectService") { serviceName: String, address: String, promise: Promise ->
            try {
                if (remoteBridge == null) {
                    throw GoBridgeNotStartedError()
                }

                remoteBridge!!.connectService(serviceName, address)
                promise.resolve(true)
            } catch (err: CodedException) {
                promise.reject(err)
            }
        }

        AsyncFunction("closeBridge") { promise: Promise ->
            try {
                if (bridgeMessenger != null) {
                    bridgeMessenger!!.close()
                    bridgeMessenger = null
                }
                if (remoteBridge != null) {
                    remoteBridge!!.close()
                    remoteBridge = null
                }
                serviceClient = null
                promise.resolve(true)
            } catch (err: CodedException) {
                promise.reject(err)
            }
        }

        AsyncFunction("invokeBridgeMethod") { method: String, b64message: String, promise: Promise ->
            try {
                if (serviceClient == null) {
                    throw GoBridgeNotStartedError()
                }

                val promiseBlock: PromiseBlock = KotlinPromiseBlock(promise)
                serviceClient!!.invokeBridgeMethodWithPromiseBlock(
                    promiseBlock,
                    method,
                    b64message
                )
            } catch (err: CodedException) {
                promise.reject(err)
            }
        }

        AsyncFunction("getProtocolAddr") { promise: Promise ->
            try {
                if (bridgeMessenger == null) {
                    throw GoBridgeNotStartedError()
                }
                promise.resolve(null)
            } catch (err: CodedException) {
                promise.reject(err)
            }
        }

        AsyncFunction("clearStorage") { promise: Promise ->
            try {
                if (rootDir != null && rootDir!!.exists()) {
                    if (!deleteRecursive(rootDir!!)) {
                        throw GoBridgeStorageError()
                    }
                }
                if (tempDir != null && tempDir!!.exists()) {
                    if (!deleteRecursive(tempDir!!)) {
                        throw GoBridgeStorageError()
                    }
                }
                promise.resolve(true)
            } catch (err: CodedException) {
                promise.reject(err)
            }
        }

        // AddressBook methods

        AsyncFunction("getDeviceCountry") { promise: Promise ->
            addressBook?.getDeviceCountry(promise);
        }


        AsyncFunction("getAllContacts") { promise: Promise ->
            addressBook?.getAllContacts(promise);
        }

        // PushToken

        AsyncFunction("request") { promise: Promise ->
            notificationService?.request(promise)
        }
    }

    private fun deleteRecursive(fileOrDirectory: File): Boolean {
        if (fileOrDirectory.isDirectory) {
            for (child in fileOrDirectory.listFiles()) {
                if (!deleteRecursive(child)) {
                    return false
                }
            }
        }
        return fileOrDirectory.delete()
    }

    fun getReactContext(): Context {
        return appContext.reactContext!!
    }

    companion object {
        var bridgeMessenger: Bridge? = null
            private set
    }
}
